model/generate.go

```diff
@@ -118,57 +117,46 @@ func ParseProjectFromJSON(data []byte) (GeneratedProject, error) {
 
 // NewVersion adds the buildvariants, tasks, and functions
 // from a generated project config to a project, and returns the previous config number.
-func (g *GeneratedProject) NewVersion() (*Project, *ParserProject, *Version, *task.Task, *projectMaps, error) {
-	// Get task, version, and project.
-	t, err := task.FindOneId(g.TaskID)
-	if err != nil {
-		return nil, nil, nil, nil, nil,
-			gimlet.ErrorResponse{StatusCode: http.StatusInternalServerError, Message: errors.Wrapf(err, "error finding task %s", g.TaskID).Error()}
-	}
-	if t == nil {
-		return nil, nil, nil, nil, nil,
-			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: fmt.Sprintf("unable to find task %s", g.TaskID)}
-	}
-	v, err := VersionFindOneId(t.Version)
-	if err != nil {
-		return nil, nil, nil, nil, nil,
-			gimlet.ErrorResponse{StatusCode: http.StatusInternalServerError, Message: errors.Wrapf(err, "error finding version %s", t.Version).Error()}
-	}
-	if v == nil {
-		return nil, nil, nil, nil, nil,
-			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: fmt.Sprintf("unable to find version %s", t.Version)}
-	}
-	p, pp, err := LoadProjectForVersion(v, t.Project, false)
-	if err != nil {
-		return nil, nil, nil, nil, nil,
-			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrapf(err, "error getting project for version %s", t.Version).Error()}
-	}
-
+func (g *GeneratedProject) NewVersion(p *Project, pp *ParserProject, v *Version) (*Project, *ParserProject, *Version, *projectMaps, error) {
```

Comment:
why do we have a public method returning a private type that isn't wrapped in an interface? 

is the *Project that's returned the same as the *Project that's passed in? can we just mutate the pointer that we're handed?

Reply:
What would you gain from not returning them? The underlying machinery doesn't mutate. You'd be generating a new thing and then assigning the pointer to the new thing. I think it's clearer to do it this way given that we're creating something new. Nothing has changed. We've just reduced the amount of work this function is doing, so now it takes an argument.